matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -301,9 +365,16 @@ private void computeTransitionProbabilities(TimeStep<GPXExtension, GPXEntry, Pat
         for (GPXExtension from : prevTimeStep.candidates) {
             for (GPXExtension to : timeStep.candidates) {
                 RoutingAlgorithm algo = algoFactory.createAlgo(queryGraph, algoOptions);
-                // System.out.println("algo " + algo.getName());
-                final Path path = algo.calcPath(from.getQueryResult().getClosestNode(),
-                        to.getQueryResult().getClosestNode());
+                // enforce heading if required:
+                if (from.isDirected()) {
+                	from.incomingVirtualEdge.setUnfavored(true);
+                }    
+                if (to.isDirected()) {
+                	// unfavor the favour virtual edge
+                	to.outgoingVirtualEdge.setUnfavored(true);
```

Comment:
Same as above. We need to make sure that the router does not take the wrong ingoing virtual edge to the virtual "to" node.